# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/WindowContext.cpp
# Commit: 812ed5275f0d
# Full Hash: 812ed5275f0db9f92944955f891df2ab68c9f8e7
# Author: Toshihito Kikuchi <tkikuchi@mozilla.com>
# Date: 2021-11-19 04:11:43
# Regressor Bug: 1734099
# File Overlap Count: 5
# Description:
#   Bug 1734099 - Never unload tabs with active RTCPeerConnection instances.  r=peterv,jesup
#   
#   This patch makes sure the Tab Unloading feature does not unload tabs that have
#   active peer connections not to disrupt browsing experience based on WebRTC peer
#   connections.
# ==============================================================================

diff -r a733d049dbda -r 812ed5275f0d docshell/base/WindowContext.cpp
--- a/docshell/base/WindowContext.cpp	Fri Nov 19 01:03:56 2021 +0200
+++ b/docshell/base/WindowContext.cpp	Thu Nov 18 22:59:12 2021 +0000
@@ -68,6 +68,12 @@
   return mBrowsingContext->mCurrentWindowContext == this;
 }
 
+bool WindowContext::HasActivePeerConnections() {
+  MOZ_ASSERT(TopWindowContext() == this,
+             "HasActivePeerConnections is set only in the top window context");
+  return Canonical()->HasActivePeerConnections();
+}
+
 bool WindowContext::IsInBFCache() {
   if (mozilla::SessionHistoryInParent()) {
     return mBrowsingContext->IsInBFCache();
@@ -280,6 +286,11 @@
   RecomputeCanExecuteScripts();
 }
 
+bool WindowContext::CanSet(FieldIndex<IDX_HasActivePeerConnections>, bool,
+                           ContentParent*) {
+  return XRE_IsParentProcess() && IsTop();
+}
+
 void WindowContext::RecomputeCanExecuteScripts(bool aApplyChanges) {
   const bool old = mCanExecuteScripts;
   if (!AllowJavascript()) {